### ISSUE TITLE
Changing validation basis

### ISSUE THREAD

**Issue by spruce** [piceaTech/ember-rapid-forms#139]
As we always had some bigger or minor issues with validations (See #13 ) and also the currently used solution no longer being maintained (See https://github.com/DockYard/ember-validations/commit/1ed7c61d6396ebf20b6953775c5b5ffdb77766f5 ). We have two options on hand and one of them is not really viable.
1. Let the user try to fix the problem themselves (See https://github.com/DockYard/ember-validations/issues/268#issuecomment-177166176 ) which is not really an option.
2. Migrate to another solution. I personally would as far as i've looked favor `ember-cp-validations` (http://offirgolan.github.io/ember-cp-validations/docs/)

What do you think about it?
@GCorbel

**Comment by GCorbel** [comment]
I aleady changed for ember-cp-validations and it works pretty well. What do we have to change ?

**Comment by spruce** [comment]
Did you just switch the validation library you use in your application and it did just work? 

I thought along the lines of creating some testcases and guides on how to properly use them. Not to much.

**Comment by GCorbel** [comment]
Yes. I'm not 100% sure but I think we just have to use [the "ds-error" validator](http://offirgolan.github.io/ember-cp-validations/docs/classes/DS%20Error.html).

**Comment by toovy** [comment]
@GCorbel @spruce I'm currently updating an old project with ember-rapid-forms and ember-validations, I had the same thought. Did you really simply drop in ember-cp-validations and it worked?

In my model (coffee-script):

    Validations = buildValidations
      newPassword: [
        validator('presence', true)
        validator('length', min: 8, max: 24)
        validator('ds-error')
      ]

Errors are not displayed as with ember-validations, here some inspection results:

    model.get('errors.length') // 0
    model.get('validations.isValid') // false
    model.get('validations.attrs.newPassword.errors') // returns a correct error object

So ember-cp-validations detects the error, but it does not seem that the DS Error is properly set on the errors object, even with using the ds-error. But I could not really find detailed documentation about the validator('ds-error'). Any thoughts?

**Comment by GCorbel** [comment]
Ok, I will check it today.

**Comment by GCorbel** [comment]
I was completly wrong and I worked **a lot** to make it working.

I did this mixin : 

```
import DS from 'ember-data';

export default Ember.Mixin.create({
  init() {
    this._super(...arguments);

    this.eachAttribute(attribute=> {
      this.addObserver(attribute, function () {
        this._copyErrors(this, attribute);
      });
    });

    this.eachRelationship(attribute=> {
      this.addObserver(attribute, function () {
        this._copyErrors(this, attribute);
      });
    });
  },

  validate() {
    return this.get('validations').validate(...arguments)
      .then(()=> {
        const modelErrors = this.get('errors');

        if (this instanceof DS.Model && !Ember.isNone(modelErrors) && Ember.canInvoke(modelErrors, 'add')) {
          this.eachAttribute(attribute=> {
            this._copyErrors(this, attribute);
          });
          this.eachRelationship(attribute=> {
            this._copyErrors(this, attribute);
          });
        }
      });
  },

  _copyErrors(model, attribute) {
    if (model.currentState.stateName === 'root.loaded.saved' || model.currentState.stateName === 'root.deleted.saved') {
      return;
    }

    const modelErrors = model.get('errors');
    const wasEmpty = modelErrors.get('isEmpty');

    if (modelErrors.get(attribute)) {
      modelErrors._remove(attribute);
    }

    const messages = this.get(`validations.attrs.${attribute}.messages`);

    if (messages && messages.length > 0) {
      messages.forEach(m=> {
        modelErrors._add(attribute, m);
      });
    }

    const isEmpty = modelErrors.get('isEmpty');

    if (wasEmpty && !isEmpty) {
      modelErrors.trigger('becameInvalid');
    } else if (!wasEmpty && isEmpty) {
      modelErrors.trigger('becameValid');
    }
  }
});
```

If I include the mixin, it works. I think it's better te contact @offirlogan to add this mixin or check if it's possible to do something in ember-cp-validations .

**Comment by spruce** [comment]
What do you think of creating a mixin inside of ember-rapid-forms so that we could offer a solution independently of ember-cp-validations?

**Comment by GCorbel** [comment]
@spruce this will not work with other addons.

**Comment by spruce** [comment]
Well I meant including a mixin like what you have done. What I want to get: Compatibilty with ember-cp-validations.

**Comment by GCorbel** [comment]
We can but it's cleaner on ember-cp-validation because this is not really related to ember-rapid-forms

**Comment by spruce** [comment]
But it is less related to ember-cp-validations. It provides a legacy mixin to support access like it was in ember-validations. And if they don't want to have a legacy layer in their codebase I could understand this. Also I think it will be faster to include the mixin in ember-rapid-forms and to try migrate it over to ember-cp-validations if possible later.

**Comment by GCorbel** [comment]
I still think it's better to do it in ember-cp-validatoin but if you think it can be useful in ERF, go ahead ! We will see what @offirlogan will do.

**Comment by GCorbel** [comment]
I discussed with @offirlogan and I will do an external addon.

**Comment by spruce** [comment]
Hey @GCorbel, 
how far have you come?

BTW: Thanks for all the effort!!!

**Comment by GCorbel** [comment]
Sorry, I'm late on this subject and I still have to do it. I will do it this week.

**Comment by spruce** [comment]
@GCorbel i couldn't wait. So I integrated it myself on a new branch. What do you think? It's a blocker for a client project of mine. I think it is a good intermediate solution.

**Comment by wmlele** [comment]
The switch to ember-cp-validations is very welcome. Just one thing, can you guys just make sure it will work even without ember-data? (currently the POC ember-cp-validations-helper depends on ED)

**Comment by spruce** [comment]
Sure. I will try to refactor later. You can review it later if you want.

**Comment by spruce** [comment]
I have just updated the PR. Could you tell me what needs to be changed to make it work without ED?
My list of thinks I think need to be changed:
- [ ] line 33: don't check currentState
- [ ] lines 49: call correct method (which would that be?)
- [ ] lines 57/59: Do these methods exist on non ember-data Models?

Could you help me with them?

**Comment by spruce** [comment]
i merged #141 just now. if you'd create a new PR with necessary changes that would be greatly appreciated.

**Comment by spruce** [comment]
I think we can close this 👍
